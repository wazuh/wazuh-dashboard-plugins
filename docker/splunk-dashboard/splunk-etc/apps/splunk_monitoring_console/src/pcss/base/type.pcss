/*===============================================================================================*/
/*  SPLUNK: TYPOGRAPHY                                                                           */
/*  Headings, body text, lists, code, and more for a versatile and durable typography system.    */
/*===============================================================================================*/

/*************************************************************************************************/
/*  BODY TEXT                                                                                    */
/*************************************************************************************************/
p {
    margin: 0 0 calc($lineHeight / 2);
}
.lead {
    margin-bottom: $lineHeight;
    font-size: calc($fontSize * 1.5);
    font-weight: 200;
    line-height: calc($lineHeight * 1.5);
}


/*************************************************************************************************/
/*  EMPHASIS & MISC                                                                              */
/*************************************************************************************************/
small               { font-size: 85%; }             /*  Ex: 14px base font * 85% = about 12px */
strong              { font-weight: bold; }
em                  { font-style: italic; }
cite                { font-style: normal; }

/*  Utility classes */
.text-left           { text-align: left; }
.text-right          { text-align: right; }
.text-center         { text-align: center; }

@define-mixin text-emphasis $class, $textColor {
    .$class {
        color: $textColor;
        a&:hover, a&:focus {
            color: rgba(60,68,77,1);
        }
    }
}

@mixin text-emphasis muted, $textGray;                /* .muted */
@mixin text-emphasis text-warning, $warningText;    /* .text-warning */
@mixin text-emphasis text-error, $errorText;        /* .text-error */
@mixin text-emphasis text-info, $infoText;          /* .text-info */
@mixin text-emphasis text-success, $successText;    /* .text-success */


/*************************************************************************************************/
/*  FONT FAMILIES                                                                                */
/*************************************************************************************************/
/* monospace for code and raw
    <p>normal</p>
    <p class="mono-space">mono space</p>
*/
.mono-space {
    font-family: $monoFontFamily;
}


/*************************************************************************************************/
/*  HEADINGS                                                                                     */
/*************************************************************************************************/
h1 {
    @mixin create-heading-extra-large;
    small { font-size: calc($fontSizeXXLarge * 0.9); }
}
h2 {
    @mixin create-heading;
    small { font-size: calc($fontSizeXLarge * 0.9); }
}
h3 {
    @mixin create-heading-medium;
    small { font-size: $fontSize; }
}
h4 {
    @mixin create-heading-small;
    font-size: $fontSize;
}
h5 {
    @mixin create-heading;
    font-size: $fontSizeSmall;
}
h6 {
    @mixin create-heading;
    font-size: $fontSizeXSmall;
}

.section-heading {
    @mixin create-section-heading;
}
.section-heading-small {
    @mixin create-section-heading $fontSizeSmall;
}


/*************************************************************************************************/
/*  PAGE HEADER                                                                                  */
/*************************************************************************************************/
.page-header {
    padding-bottom: calc(($lineHeight / 2) - 1px);
    margin: $lineHeight 0 calc($lineHeight * 1.5);
    border-bottom: $borderLight;
}


/*************************************************************************************************/
/*  LISTS                                                                                        */
/*************************************************************************************************/
/*  Unordered and Ordered lists */
ul, ol {
    padding: 0;
    margin: 0 0 calc($lineHeight / 2) 25px;

    ul, ol {
        margin-bottom: 0;
    }
    /*  Remove default list styles */
    &.unstyled {
        margin-left: 0;
        list-style: none;
    }
    /*  Single-line list items */
    &.inline {
        margin-left: 0;
        list-style: none;

        > li {
            display: inline-block;
            padding-left: $spacingQuarter;
            padding-right: $spacingQuarter;
        }
    }
}
li {
    line-height: $lineHeight;
}

/*  Description Lists */
dl {
    margin-bottom: $lineHeight;
}
dt,
dd {
    line-height: $lineHeight;
}
dt {
    font-weight: bold;
}
dd {
    margin-left: calc($lineHeight / 2);
}

/*  Horizontal layout (like forms) */
.dl-horizontal {
    @mixin clearfix; /*  Ensure dl clears floats if empty dd elements present */
    dt {
        float: left;
        width: calc($horizontalComponentOffset - 20px);
        clear: left;
        text-align: right;
        overflow: hidden;
        text-overflow: ellipsis;
        white-space: nowrap;
    }
    dd {
        margin-left: $horizontalComponentOffset;
    }
}

/*************************************************************************************************/
/*  MISC                                                                                         */
/*************************************************************************************************/
/*  Horizontal rules */
hr {
    margin: $lineHeight 0;
    border: 0;
    border-top: $borderLight;
    border-bottom: $borderWhite;
}

/*  Abbreviations and acronyms */
abbr[title],
    /*  Added data-* attribute to help out our tooltip plugin, per https://github.com/twitter/bootstrap/issues/5257 */
abbr[data-original-title] {
    cursor: help;
    border-bottom: 1px dotted $gray60;
}

abbr.initialism {
    font-size: 90%;
    text-transform: uppercase;
}

/*  Blockquotes */
blockquote {
    padding: 0 0 0 $spacingThreeQuarters;
    margin: 0 0 $lineHeight;
    border-left: 5px solid $gray92;
    p {
        margin-bottom: 0;
        font-size: calc($fontSize * 1.25);
        font-weight: 300;
        line-height: 1.25;
    }
    small {
        display: block;
        line-height: $lineHeight;
        color: $gray60;
        &:before {
            content: '\2014 \00A0';
        }
    }
    /*  Float right with text-align: right */
    &.pull-right {
        float: right;
        padding-right: $spacingThreeQuarters;
        padding-left: 0;
        border-right: 5px solid $gray92;
        border-left: 0;
        p,
        small {
            text-align: right;
        }
        small {
            &:before {
                content: '';
            }
            &:after {
                content: '\00A0 \2014';
            }
        }
    }
}

/*  Quotes */
q:before,
q:after,
blockquote:before,
blockquote:after {
    content: "";
}

/*  Addresses */
address {
    display: block;
    margin-bottom: $lineHeight;
    font-style: normal;
    line-height: $lineHeight;
}
