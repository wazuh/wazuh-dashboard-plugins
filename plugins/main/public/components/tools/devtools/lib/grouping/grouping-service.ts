import $ from 'jquery';
import type { EditorGroup, EditorLike } from '../types/editor';
import { REQUEST_SPLIT_REGEX } from '../constants/regex';
import { ErrorService } from '../services/error-service';
import { EDITOR_HIGHLIGHT_CLASS } from '../constants/ui';
import { resolveDocsUrl } from './services/docs-link-resolver';
import type { EditorUIController } from './services/ui-controller';
import { JQueryEditorUIController } from './services/ui-controller';
import type { CurrentStateStore } from './services/state-adapter';
import { DefaultCurrentStateStore } from './services/state-adapter';
import type { JsonLinter } from './services/json-linter';
import { DefaultJsonLinter } from './services/json-linter';

/**
 * Service that encapsulates request grouping logic and related UI hooks.
 * All dependencies are injected for testability; defaults are provided.
 */
export class GroupingService {
  constructor(
    private ui: EditorUIController = new JQueryEditorUIController(),
    private errors = new ErrorService(),
    private state: CurrentStateStore = new DefaultCurrentStateStore(),
    private json: JsonLinter = new DefaultJsonLinter(),
  ) {}

  /**
   * Split the current buffer into request groups (method + path + optional JSON body).
   */
  parseGroups(editor: EditorLike): EditorGroup[] {
    try {
      const currentState = String(editor.getValue());
      this.state.setCurrentDevTools(currentState);

      const tmpgroups: EditorGroup[] = [];
      const splitted = currentState
        .split(REQUEST_SPLIT_REGEX)
        .filter(item => item.replace(/\s/g, '').length);

      let start = 0;
      let end = 0;
      let starts: number[] = [];
      const slen = splitted.length;
      for (let i = 0; i < slen; i++) {
        let tmp: string[] = splitted[i].split('\n');
        if (Array.isArray(tmp)) tmp = tmp.filter(item => !item.startsWith('#'));
        const cursor = editor.getSearchCursor?.(splitted[i], null, {
          multiline: true,
        });

        if (cursor && cursor.findNext()) start = cursor.from().line;
        else return [];

        if (tmp.length) {
          while (
            editor.getLine?.(cursor.from().line) !== tmp[0] &&
            cursor.findNext()
          ) {
            start = cursor.from().line;
          }
          while (starts.includes(start) && cursor.findNext()) {
            start = cursor.from().line;
          }
        }
        starts.push(start);

        end = start + tmp.length;

        const tmpRequestText = tmp[0];
        let tmpRequestTextJson = '';

        const tmplen = tmp.length;
        for (let j = 1; j < tmplen; ++j) {
          if (!!tmp[j] && !tmp[j].startsWith('#')) {
            tmpRequestTextJson += tmp[j];
          }
        }

        if (tmpRequestTextJson && typeof tmpRequestTextJson === 'string') {
          let rtjlen = tmp.length;
          const isClosingLine = (s: string) => {
            const t = s.trim();
            // Accept '}' or ']' optionally followed by a comma
            return /^(\}|\])\s*,?$/.test(t);
          };
          while (rtjlen--) {
            if (isClosingLine(tmp[rtjlen])) break;
            else end -= 1;
          }
        }

        if (!tmpRequestTextJson && tmp.length > 1) {
          tmp = [tmp[0]];
          end = start + 1;
        }

        if (i === slen - 1 && !tmpRequestTextJson) {
          if (tmp.length > 1) end -= tmp.length - 1;
        }

        end--;

        tmpgroups.push({
          requestText: tmpRequestText,
          requestTextJson: tmpRequestTextJson,
          start,
          end,
        });
      }
      starts = [];
      return tmpgroups;
    } catch (error: unknown) {
      this.errors.log({ context: 'parseGroups', error });
      return [];
    }
  }

  /**
   * Choose which group is active based on the cursor position and place action buttons.
   */
  selectActiveGroup(
    editor: EditorLike,
    firstTime = false,
    groups: EditorGroup[] = [],
  ): EditorGroup | null {
    try {
      const selection = editor.getCursor();
      const validGroups = groups.filter(item => item.requestText);
      let desiredGroup: EditorGroup[] = firstTime
        ? validGroups
        : validGroups.filter(
            item =>
              item.requestText &&
              item.end >= selection.line &&
              item.start <= selection.line,
          );

      if (!firstTime && desiredGroup.length === 0) {
        try {
          const wrapperEl =
            editor.getWrapperElement?.() || editor.display?.wrapper || null;
          const $wrapper = wrapperEl ? $(wrapperEl) : null;
          const wrapperTop = $wrapper?.offset()?.top ?? 0;
          const wrapperHeight = $wrapper?.outerHeight?.() ?? 0;
          const wrapperBottom = wrapperTop + wrapperHeight;

          const firstVisible = validGroups.find(g => {
            try {
              const c = editor.cursorCoords({ line: g.start, ch: 0 });
              const lineTop = c.top + 1;
              return lineTop >= wrapperTop && lineTop <= wrapperBottom;
            } catch {
              return false;
            }
          });

          if (firstVisible) {
            desiredGroup = [firstVisible];
          } else if (validGroups.length) {
            desiredGroup = [validGroups[0]];
          }
        } catch {}

        if (!desiredGroup.length) {
          this.ui.hidePlay();
          this.ui.hideDocs();
          return null;
        }
      }

      let cords: { top: number };
      try {
        cords = editor.cursorCoords({ line: desiredGroup[0]?.start, ch: 0 });
      } catch {
        this.ui.hidePlay();
        this.ui.hideDocs();
        return null;
      }

      if (desiredGroup[0]) {
        try {
          const wrapperEl =
            editor.getWrapperElement?.() || editor.display?.wrapper || null;
          const $wrapper = wrapperEl ? $(wrapperEl) : null;
          const wrapperTop = $wrapper?.offset()?.top ?? 0;
          const wrapperHeight = $wrapper?.outerHeight?.() ?? 0;
          const wrapperBottom = wrapperTop + wrapperHeight;

          const lineTop = cords.top + 1;

          const isVisible = lineTop >= wrapperTop && lineTop <= wrapperBottom;

          if (!isVisible) {
            this.ui.hidePlay();
            this.ui.hideDocs();
          } else {
            this.ui.showPlay();
            this.ui.showDocs(
              resolveDocsUrl(editor, desiredGroup[0].requestText) || '',
            );
            this.ui.setButtonsTop(lineTop);
          }
        } catch {
          this.ui.hidePlay();
          this.ui.hideDocs();
        }
      }

      if (firstTime) this.highlightGroup(editor, desiredGroup[0]);

      // Validate and toggle docs. Keep the play button always available
      // so users can send requests to any endpoint, even if we don't
      // have a documentation URL for it.
      if (desiredGroup[0]) {
        const url = resolveDocsUrl(editor, desiredGroup[0].requestText);
        if (url) {
          this.ui.showDocs(url);
          this.ui.showPlay();
        } else {
          this.ui.hideDocs();
          // Do not hide the play button when docs URL is missing
          this.ui.showPlay();
        }
      }

      return desiredGroup[0] || null;
    } catch (error: unknown) {
      this.ui.hidePlay();
      this.ui.hideDocs();
      this.errors.log({
        context: 'selectActiveGroup',
        error,
        level: undefined,
      });
      return null;
    }
  }

  /**
   * Mark the given group as active by highlighting its lines.
   */
  highlightGroup(editor: EditorLike, group?: EditorGroup) {
    if (!editor.__highlightedLines) {
      editor.__highlightedLines = [] as number[];
    }

    for (const lineNo of editor.__highlightedLines) {
      editor.removeLineClass?.(lineNo, 'background', EDITOR_HIGHLIGHT_CLASS);
    }
    editor.__highlightedLines = [];

    if (!group) return;

    const add = (ln: number) => {
      editor.addLineClass?.(ln, 'background', EDITOR_HIGHLIGHT_CLASS);
      editor.__highlightedLines!.push(ln);
    };

    if (
      !group.requestTextJson ||
      (group.requestText.includes('{') && group.requestText.includes('}'))
    ) {
      add(group.start);
      return;
    }

    for (let i = group.start; i <= group.end; i++) add(i);
  }

  /**
   * Validate JSON structure for all groups and render error widgets.
   * Returns the list of requestText identifiers that failed.
   */
  validateJson(editor: EditorLike, groups: EditorGroup[] = []): string[] {
    const affectedGroups: string[] = [];
    for (const widget of editor.__widgets || []) {
      editor.removeLineWidget?.(widget.widget);
    }
    editor.__widgets = [];
    for (const item of groups) {
      if (item.requestTextJson) {
        try {
          this.json.parse(item.requestTextJson);
        } catch (error: unknown) {
          affectedGroups.push(item.requestText);
          const msg = window.document.createElement('div');
          msg.id = String(new Date().getTime() / 1000);
          const icon = msg.appendChild(window.document.createElement('div'));

          icon.className = 'lint-error-icon';
          icon.id = String(new Date().getTime() / 1000);
          icon.onmouseover = () => {
            const advice = msg.appendChild(
              window.document.createElement('span'),
            );
            advice.id = String(new Date().getTime() / 1000);
            advice.innerText =
              (error?.message as string) || 'Error parsing query';
            advice.className = 'lint-block-wz';
          };

          icon.onmouseleave = () => {
            if (msg.lastChild) msg.removeChild(msg.lastChild as Node);
          };

          editor.__widgets.push({
            start: item.start,
            widget: editor.addLineWidget?.(item.start, msg, {
              coverGutter: false,
              noHScroll: true,
            }),
          });
        }
      }
    }
    return affectedGroups;
  }
}
