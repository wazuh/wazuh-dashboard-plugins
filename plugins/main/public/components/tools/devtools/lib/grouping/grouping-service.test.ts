/**
 * Tests for GroupingService covering parsing, selection, highlighting, and JSON validation.
 */
import type { EditorGroup, EditorLike } from '../types/editor';
import { GroupingService } from './grouping-service';

// Mock docs resolver so we control URL behavior
jest.mock('./services/docs-link-resolver', () => ({
  resolveDocsUrl: jest.fn((_: any, requestText: string) =>
    requestText?.includes('/docs') ? 'https://docs.example/api' : null,
  ),
}));

// Minimal jQuery mock for wrapper metrics used by selectActiveGroup
jest.mock('jquery', () => {
  const $ = jest.fn((el: any) => ({
    offset: () => ({ top: 0 }),
    outerHeight: () => 1000,
    is: () => true,
    show: jest.fn(),
    hide: jest.fn(),
    attr: jest.fn(),
  }));
  return $.default || $;
});

function buildEditor(content: string): EditorLike & { __lines: string[] } {
  const lines = content.split('\n');
  const selection = { line: 0, ch: 0 };

  const editor: any = {
    __lines: lines,
    getValue: () => content,
    getCursor: () => ({ ...selection }),
    setCursor: (pos: any) => {
      selection.line = pos.line;
      selection.ch = pos.ch || 0;
    },
    getLine: (n: number) => lines[n] ?? '',
    cursorCoords: ({ line }: any) => ({ top: line * 20 }),
    getWrapperElement: () => document.createElement('div'),
    getSearchCursor: (query: string) => {
      // Find by matching the first line of the query
      const first = (query || '').split('\n')[0];
      const idx = lines.findIndex(l => l === first);
      let used = false;
      return {
        findNext: () => {
          if (used) return false;
          used = true;
          return idx >= 0;
        },
        from: () => ({ line: idx >= 0 ? idx : 0, ch: 0 }),
      };
    },
    addLineClass: jest.fn(),
    removeLineClass: jest.fn(),
    addLineWidget: jest.fn((_line: number, node: any) => ({ node })),
    removeLineWidget: jest.fn(),
  };
  return editor as any;
}

describe('GroupingService.parseGroups', () => {
  it('splits text into groups with correct ranges and JSON bodies', () => {
    const content = [
      'GET /agents',
      '',
      'POST /agents',
      '{',
      '  "foo": "bar"',
      '}',
      '# ignore this comment',
      'DELETE /agents/123?pretty=true',
      '',
      'PUT /agents/001/docs',
    ].join('\n');

    const editor = buildEditor(content);

    const mockState = { setCurrentDevTools: jest.fn() };
    const mockJson = { parse: jest.fn() };
    const svc = new GroupingService(
      // ui
      {
        showPlay: jest.fn(),
        hidePlay: jest.fn(),
        showDocs: jest.fn(),
        hideDocs: jest.fn(),
        setButtonsTop: jest.fn(),
      } as any,
      // errors
      { log: jest.fn() } as any,
      // state
      mockState as any,
      // json linter
      mockJson as any,
    );

    const groups = svc.parseGroups(editor);

    // State persistence invoked
    expect(mockState.setCurrentDevTools).toHaveBeenCalledWith(content);

    expect(groups).toHaveLength(4);

    const [g0, g1, g2, g3] = groups;

    expect(g0).toMatchObject<Partial<EditorGroup>>({
      requestText: 'GET /agents',
      requestTextJson: '',
      start: 0,
      end: 0,
    });

    expect(g1).toMatchObject<Partial<EditorGroup>>({
      requestText: 'POST /agents',
      requestTextJson: '{  "foo": "bar"}',
      start: 2,
      end: 5,
    });

    expect(g2).toMatchObject<Partial<EditorGroup>>({
      requestText: 'DELETE /agents/123?pretty=true',
      requestTextJson: '',
      start: 7,
      end: 7,
    });

    expect(g3).toMatchObject<Partial<EditorGroup>>({
      requestText: 'PUT /agents/001/docs',
      requestTextJson: '',
      start: 9,
      end: 9,
    });

    // Snapshot full structure
    expect(groups).toMatchSnapshot();
  });

  it('returns empty array if search cursor cannot find a block', () => {
    const editor = buildEditor('GET /agents');
    editor.getSearchCursor = () => ({
      findNext: () => false,
      from: () => ({ line: 0, ch: 0 }),
    });

    const svc = new GroupingService();
    const groups = svc.parseGroups(editor);
    expect(groups).toEqual([]);
  });
});

describe('GroupingService.highlightGroup', () => {
  it('clears previous highlights and highlights a single-line group', () => {
    const editor = buildEditor('GET /agents');
    (editor as any).__highlightedLines = [10, 11];

    const svc = new GroupingService();
    svc.highlightGroup(editor, {
      requestText: 'GET /agents',
      requestTextJson: '',
      start: 0,
      end: 0,
    });

    // Previous highlights removed
    expect(editor.removeLineClass).toHaveBeenCalledWith(
      10,
      'background',
      expect.any(String),
    );
    expect(editor.removeLineClass).toHaveBeenCalledWith(
      11,
      'background',
      expect.any(String),
    );

    // New highlight applied only to start line
    expect(editor.addLineClass).toHaveBeenCalledWith(
      0,
      'background',
      expect.any(String),
    );
  });

  it('highlights all lines in a multi-line JSON group', () => {
    const editor = buildEditor('POST /agents\n{\n  "a": 1\n}');
    const svc = new GroupingService();
    svc.highlightGroup(editor, {
      requestText: 'POST /agents',
      requestTextJson: '{"a":1}',
      start: 5,
      end: 8,
    });

    expect(editor.addLineClass).toHaveBeenCalledTimes(4); // 5,6,7,8
    expect(editor.addLineClass).toHaveBeenCalledWith(
      5,
      'background',
      expect.any(String),
    );
    expect(editor.addLineClass).toHaveBeenCalledWith(
      8,
      'background',
      expect.any(String),
    );
  });
});

describe('GroupingService.selectActiveGroup', () => {
  function makeUI() {
    return {
      showPlay: jest.fn(),
      hidePlay: jest.fn(),
      showDocs: jest.fn(),
      hideDocs: jest.fn(),
      setButtonsTop: jest.fn(),
    };
  }

  it('selects group by cursor position and shows buttons when visible', () => {
    const content = ['GET /agents', '', 'GET /agents/docs'].join('\n');

    const editor = buildEditor(content);
    const ui = makeUI();
    const svc = new GroupingService(ui as any, { log: jest.fn() } as any);

    const groups: EditorGroup[] = [
      { requestText: 'GET /agents', requestTextJson: '', start: 0, end: 0 },
      {
        requestText: 'GET /agents/docs',
        requestTextJson: '',
        start: 2,
        end: 2,
      },
    ];

    // Put cursor on line 2 (docs request), and ensure visible
    (editor as any).setCursor({ line: 2, ch: 0 });

    const selected = svc.selectActiveGroup(editor, false, groups);
    expect(selected).toEqual(groups[1]);

    // Docs URL resolver mocked only returns for paths including '/docs'
    expect(ui.showDocs).toHaveBeenCalledWith('https://docs.example/api');
    expect(ui.showPlay).toHaveBeenCalled();
    expect(ui.setButtonsTop).toHaveBeenCalledWith(2 * 20 + 1);
  });

  it('falls back to first visible group when cursor not inside any range', () => {
    const content = ['GET /a/docs', 'GET /b'].join('\n');
    const editor = buildEditor(content);
    const ui = makeUI();
    const svc = new GroupingService(ui as any, { log: jest.fn() } as any);

    const groups: EditorGroup[] = [
      { requestText: 'GET /a/docs', requestTextJson: '', start: 0, end: 0 },
      { requestText: 'GET /b', requestTextJson: '', start: 1, end: 1 },
    ];

    // Cursor below all ranges
    (editor as any).setCursor({ line: 5, ch: 0 });

    const selected = svc.selectActiveGroup(editor, false, groups);
    expect(selected).toEqual(groups[0]);
    expect(ui.showDocs).toHaveBeenCalledWith('https://docs.example/api');
    expect(ui.showPlay).toHaveBeenCalled();
  });

  it('keeps play button visible even if no docs URL is available', () => {
    const content = ['GET /a', 'GET /b'].join('\n');
    const editor = buildEditor(content);
    const ui = makeUI();
    const svc = new GroupingService(ui as any, { log: jest.fn() } as any);

    const groups: EditorGroup[] = [
      { requestText: 'GET /a', requestTextJson: '', start: 0, end: 0 },
      { requestText: 'GET /b', requestTextJson: '', start: 1, end: 1 },
    ];

    (editor as any).setCursor({ line: 1, ch: 0 });
    const selected = svc.selectActiveGroup(editor, false, groups);
    expect(selected).toEqual(groups[1]);
    // No docs URL hides docs link but keeps play visible
    expect(ui.hideDocs).toHaveBeenCalled();
    expect(ui.showPlay).toHaveBeenCalled();
  });

  it('calls highlight on firstTime=true', () => {
    const content = 'GET /a/docs';
    const editor = buildEditor(content);
    const ui = makeUI();
    const svc = new GroupingService(ui as any, { log: jest.fn() } as any);
    const spy = jest.spyOn(svc, 'highlightGroup').mockImplementation(() => {});

    const groups: EditorGroup[] = [
      { requestText: 'GET /a/docs', requestTextJson: '', start: 0, end: 0 },
    ];
    (editor as any).setCursor({ line: 0, ch: 0 });
    svc.selectActiveGroup(editor, true, groups);
    expect(spy).toHaveBeenCalledWith(editor, groups[0]);
  });

  it('handles exceptions by hiding buttons and returning null', () => {
    const editor = buildEditor('GET /a');
    // Force error
    (editor as any).getCursor = () => {
      throw new Error('boom');
    };
    const ui = makeUI();
    const errors = { log: jest.fn() };
    const svc = new GroupingService(ui as any, errors as any);

    const res = svc.selectActiveGroup(editor, false, [
      { requestText: 'GET /a', requestTextJson: '', start: 0, end: 0 },
    ]);
    expect(res).toBeNull();
    expect(ui.hidePlay).toHaveBeenCalled();
    expect(ui.hideDocs).toHaveBeenCalled();
    expect(errors.log).toHaveBeenCalled();
  });
});

describe('GroupingService.validateJson', () => {
  it('lints JSON, adds error widgets and returns affected requestTexts', () => {
    const editor = buildEditor('POST /a');
    // Pre-existing widgets should be removed
    (editor as any).__widgets = [{ start: 1, widget: { id: 'old' } }];

    const json = {
      parse: jest.fn((raw: string) => {
        if (raw.includes('bad')) throw new Error('Unexpected token');
      }),
    };

    const svc = new GroupingService(
      {
        showPlay: jest.fn(),
        hidePlay: jest.fn(),
        showDocs: jest.fn(),
        hideDocs: jest.fn(),
        setButtonsTop: jest.fn(),
      } as any,
      { log: jest.fn() } as any,
      { setCurrentDevTools: jest.fn() } as any,
      json as any,
    );

    const groups: EditorGroup[] = [
      {
        requestText: 'POST /ok',
        requestTextJson: '{"good": true}',
        start: 0,
        end: 3,
      },
      {
        requestText: 'POST /fail',
        requestTextJson: '{bad json}',
        start: 4,
        end: 8,
      },
      { requestText: 'GET /noop', requestTextJson: '', start: 9, end: 9 },
    ];

    const failed = svc.validateJson(editor, groups);
    expect(failed).toEqual(['POST /fail']);
    expect(editor.removeLineWidget).toHaveBeenCalledWith({ id: 'old' });
    expect(Array.isArray((editor as any).__widgets)).toBe(true);
    expect((editor as any).__widgets[0].start).toBe(4);
    expect(editor.addLineWidget).toHaveBeenCalled();
  });
});
