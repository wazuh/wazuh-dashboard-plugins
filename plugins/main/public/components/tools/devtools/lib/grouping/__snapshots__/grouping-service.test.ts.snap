// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupingService.parseGroups splits text into groups with correct ranges and JSON bodies 1`] = `
Array [
  Object {
    "end": 0,
    "requestText": "GET /agents",
    "requestTextJson": "",
    "start": 0,
  },
  Object {
    "end": 5,
    "requestText": "POST /agents",
    "requestTextJson": "{  \\"foo\\": \\"bar\\"}",
    "start": 2,
  },
  Object {
    "end": 7,
    "requestText": "DELETE /agents/123?pretty=true",
    "requestTextJson": "",
    "start": 7,
  },
  Object {
    "end": 9,
    "requestText": "PUT /agents/001/docs",
    "requestTextJson": "",
    "start": 9,
  },
]
`;
